# North coast kelp data conversion - draft 2

Resources:
- [Wiki for pyodbc](https://github.com/mkleehammer/pyodbc/wiki)

In [1]:
## Imports

import pandas as pd
import numpy as np
import csv, pyodbc
import pickle
import datetime

In [2]:
## Ensure my general functions for the MPA data integration project can be imported, and import them

import sys
sys.path.insert(0, "C:\\Users\\dianalg\\PycharmProjects\\PythonScripts\\MPA data integration")

import WoRMS # functions for querying WoRMS REST API

## Connect to db and retrieve data

Note that (as described in the Wiki) the original database filename did not work because it included underscores. I renamed 'Abalone_DiveSurveys_EH_06242020.mdb' to 'AbaloneDiveSurveys-06242020.mdb' to fix this problem

In [3]:
## Check available access driver

[x for x in pyodbc.drivers() if x.startswith('Microsoft Access Driver')]

['Microsoft Access Driver (*.mdb, *.accdb)']

In [4]:
## Function for connecting to db

def get_data_from_db(tbl_name):
    """Connects to .mdb file and extracts data from tbl_name. Saves data in a .csv file and column names in a .txt file."""
    
    # Connect
    conn_str = (
        r'DRIVER={Microsoft Access Driver (*.mdb, *.accdb)};'
        r'DBQ=C:\Users\dianalg\PycharmProjects\PythonScripts\MPA data integration\North coast kelp\AbaloneDiveSurveys-06242020.mdb;'
        )
    cnxn = pyodbc.connect(conn_str)
    crsr = cnxn.cursor()
    
    # Get column names
    cols = [row.column_name for row in crsr.columns(table=tbl_name)]
    
    # Get rows
    SQL = 'SELECT * FROM ' + tbl_name + ';'
    rows = crsr.execute(SQL).fetchall()
    
    # Close connection
    crsr.close()
    cnxn.close()
    
    # Save data
    with open(tbl_name + '.csv', 'w') as file:
        csv_writer = csv.writer(file) 
        csv_writer.writerows(rows)
        
    # Save column names
    with open(tbl_name + '_cols.data', 'wb') as file:
        pickle.dump(cols, file)

**Note** that the table names in this database are:
- Paste Errors
- Switchboard Items
- Tbl_Diver_Lookup
- Tbl_Format_Lookup
- Tbl_New_size
- Tbl_SppSeen
- Tbl_Survey_Lookup
- tblCounts
- tblHabitat
- tblNewGrowth
- tblQuadrat
- tblSite
- TblSpecies
- tblSubstrate
- tblSurvey
- tblSwimGroups

In [5]:
## Extract data from .mdb file

tables = ['tblSite', 'tblSurvey', 'tblCounts', 'tblSpecies', 'Tbl_New_size', 'tblSubstrate', 'tblHabitat']

for table in tables:
    get_data_from_db(table)

## Load data

In [3]:
## Function to load data

def load_table(tbl_name):
    """Takes tbl_name (a string) and loads saved data from that table."""
    
    # Get filenames
    col_name = tbl_name + '_cols.data'
    data_name = tbl_name + '.csv'
    
    # Retrieve column names
    with open(col_name, 'rb') as file:
        cols = pickle.load(file)
        
    # Load data
    data = pd.read_csv(data_name, header=None, names=cols, encoding='ANSI')
    return(data)

In [4]:
## Load data

site = load_table('tblSite')
survey = load_table('tblSurvey')
count = load_table('tblCounts')
species = load_table('tblSpecies')
size = load_table('Tbl_New_size')
substrate = load_table('tblSubstrate')
habitat = load_table('tblHabitat')

In [5]:
## Tidy survey table

# Select relevant columns
sur = survey[[
    'Survey_ID', 
    'SiteID', 
    'Survey_Num', 
    'SURVEY', 
    'DATE', 
    'Avg Depth',
    'Min_DEPTH', 
    'Max_Depth',
    'SLAT',
    'SLONG',
    'ELAT',
    'ELONG',
    'SLAT_DD',
    'SLONG_DD',
    'ELAT_DD',
    'ELONG_DD',
    'COMMENTS',
]]

# Filter survey type as instructed by Laura
print(sur.shape)
surveys_to_keep = [
    'Transect - 30m (Rapid Emergent)',
    'Transect - 30mx2m (Emergent)',
    'Transect - 30m (Emergent)',
]
sur = sur[sur['SURVEY'].isin(surveys_to_keep)]
print(sur.shape)

# View
sur.head()

(3904, 17)
(2943, 17)


Survey_ID SiteID   Survey_Num                           SURVEY  \
0       4790     FR    FR18-D5-1  Transect - 30m (Rapid Emergent)   
1       4919    ALB  ALB18-A02-1  Transect - 30m (Rapid Emergent)   
2       4920    ALB  ALB18-A02-2  Transect - 30m (Rapid Emergent)   
3       4915    ALB   ALB18-A1-1  Transect - 30m (Rapid Emergent)   
4       4916    ALB   ALB18-A1-2  Transect - 30m (Rapid Emergent)   

                  DATE  Avg Depth  Min_DEPTH  Max_Depth  SLAT  SLONG  ELAT  \
0  2018-07-12 00:00:00        NaN       48.0        NaN   NaN    NaN   NaN   
1  2018-08-30 00:00:00        NaN        7.0       10.0   NaN    NaN   NaN   
2  2018-08-30 00:00:00        NaN       10.0       14.0   NaN    NaN   NaN   
3  2018-08-30 00:00:00        NaN       12.0       16.0   NaN    NaN   NaN   
4  2018-08-30 00:00:00        NaN        8.0        8.0   NaN    NaN   NaN   

   ELONG  SLAT_DD  SLONG_DD  ELAT_DD  ELONG_DD  \
0    NaN      0.0       0.0      0.0       0.0   
1    NaN      0.0       0.0      0.0       0.0   
2    NaN      0.0       0.0      0.0       0.0   
3    NaN      0.0       0.0      0.0       0.0   
4    NaN      0.0       0.0      0.0       0.0   

                                            COMMENTS  
0  RAPID EMERGENT SURVEY. D5: 1 OUT OF 2 TRANSECT...  
1  RAPID EMERGENT SITE A02: 1 OF 2 TRANSECTS. HEA...  
2  RAPID EMERGENT SITE A02: 2 OF 2 TRANSECTS. HEA...  
3  RAPID EMERGENT SITE A1: 1 OF 4 TRANSECTS. HEAD...  
4  RAPID EMERGENT SITE A1: 2 OF 4 TRANSECTS. HEAD...

**Some notes about coordinates in this data set:**

1. I have assumed we don't ever want to use the "old" coordinates.
2. I have no idea what the distinction between S and E is.
3. There are fewer missing values in the SLAT, SLONG and ELAT, ELONG fields than in the corresponding decimal degree columns. But these are in some weird coordinate system. 
4. Generally, the LAT and LONG values come in pairs (one is not missing while the other is present). The exception is for ELAT and ELONG, where 72 ELATs exist without an ELONG. My guess here is that SLAT was accidentally entered into the ELAT column. These column values are identical for the 72 records, with one exception.
5. Generally, the E coordinates are present far, far less than the S coordinates. **So I'm going to use the S coordinates for now.**
6. SLAT and SLONG exist for all records where SLAT_DD and SLONG_DD exist. However, they also exist for an additional 910 rows. I assume some sort of conversion could be used to obtain DD units for these records.

```python
# Number of non-zero and non-missing values for all coordinate columns
num_na = sur.isna().sum()
num_na_or_not_zero = sur.astype(bool).sum()
num_na_or_not_zero - num_na

# There are 72 rows where ELAT exists but ELONG does not
s = sur[(sur['ELAT'].isna() == False) & (sur['ELAT'] != 0)]
t = s[s['ELONG'].isna() == True]
t

# For all but one of these rows, ELAT == SLAT
t[t['SLAT'] != t['ELAT']]

# SLAT and SLONG exist in 910 records where SLAT_DD and SLONG_DD do not
u = sur[(sur['SLAT'].isna() == False) & (sur['SLAT'] != 0 ) & ((sur['SLAT_DD'].isna() == True) | (sur['SLAT_DD'] == 0))]
u

# SLAT and SLONG also exist for all records where SLAT_DD and SLONG_DD exist
v = sur[(sur['SLAT_DD'].isna() == False) & (sur['SLAT_DD'] != 0)]
v[(v['SLAT'].isna() == True) | (v['SLAT'] == 0)]
```

**I need to put together a data table showing which coordinates I have, which need to be converted, and which are missing alltogether to share with Laura. For now, I'm just going to use the SLAT_DD and SLONG_DD fields.**

In [6]:
## Put together tables of missing coordinates for Laura

# table = sur[['Survey_ID', 'Survey_Num', 'SLAT', 'SLONG', 'SLAT_DD', 'SLONG_DD']].copy()
# no_coords = table[(table['SLAT'].isna() == True) | (table['SLAT'] == 0)] # 1416 records
# coords_to_convert = table[(table['SLAT'].isna() == False) & (table['SLAT'] != 0) & ((table['SLAT_DD'].isna() == True) | (table['SLAT_DD'] == 0))] # 910 records
# no_coords.to_csv('no_coords.csv')
# coords_to_convert.to_csv('coords_to_convert.csv')

## Convert

This submission will require an event, occurrence, and MoF file. As with the RCCA and PISCO data, we'll treat transects as events and observations of organism as occurrences. Measurements will include organism sizes and substrate information.

The **event** file should contain: eventID, eventDate, datasetID, locality, localityRemarks (if needed), countryCode, decimalLat, decimalLon, coordinateUncertaintyInMeters, minimumDepthInMeters, maximumDepthInMeters, samplingProtocol, samplingEffort.

The **occurrence** file should contain: occurrenceID, scienficName, scientificNameID, taxonID, nameAccordingTo, occurrenceStatus, basisOfRecord, identificationQualifier (if needed), occurrenceRemarks (if needed), individualCount (or organismQuantity and organismQuantityType).

### eventID

Each record in the survey table has a unique Survey_Num, so I'll use that as an eventID. Laura says these should be composed of a two letter site code + a two number year + a dash + transect number. Transects also should generally have a depth level associated with them (A (shallow 0-15 ft), B (16-30 ft), C (31-45 ft), or D (46-60 ft)). There will be some inconsistencies in Survey_Num format, especially for older years. 

In [7]:
## eventID 

event = pd.DataFrame({'eventID':sur['Survey_Num']})
print(event.shape)
event.head()

(2943, 1)


eventID
0    FR18-D5-1
1  ALB18-A02-1
2  ALB18-A02-2
3   ALB18-A1-1
4   ALB18-A1-2

In [8]:
## eventDate

event['eventDate'] = sur['DATE']

# Drop rows where date is missing
print(event.shape)
event.dropna(subset=['eventDate'], inplace=True)
print(event.shape)

# Format
eventDate = [datetime.datetime.strptime(dt, '%Y-%m-%d %H:%M:%S').date().isoformat() for dt in event['eventDate']]
event['eventDate'] = eventDate
event.head()

(2943, 2)
(2941, 2)


eventID   eventDate
0    FR18-D5-1  2018-07-12
1  ALB18-A02-1  2018-08-30
2  ALB18-A02-2  2018-08-30
3   ALB18-A1-1  2018-08-30
4   ALB18-A1-2  2018-08-30

**Note** that there are two surveys with missing dates: OC18-B1-4 and OC18-B8-1. **I've dropped these for now.**

There are 289 unique survey dates in this data set. As expected, the earliest of these is 1971-09-01 and the latest of these is 2019-08-16. 

In [9]:
## datasetID

event['datasetID'] = 'North coast kelp emergent transects'
event.head()

eventID   eventDate                            datasetID
0    FR18-D5-1  2018-07-12  North coast kelp emergent transects
1  ALB18-A02-1  2018-08-30  North coast kelp emergent transects
2  ALB18-A02-2  2018-08-30  North coast kelp emergent transects
3   ALB18-A1-1  2018-08-30  North coast kelp emergent transects
4   ALB18-A1-2  2018-08-30  North coast kelp emergent transects

In [10]:
## locality

event['SiteID'] = sur['SiteID']
event = event.merge(site, how='left', on='SiteID')
event.rename(columns={'SITE':'locality'}, inplace=True)

## ------ TODO: Drop SiteID, but for now leave it in in case I need to merge on it again.

print(event.shape)
event.head()

(2941, 5)


eventID   eventDate                            datasetID SiteID  \
0    FR18-D5-1  2018-07-12  North coast kelp emergent transects     FR   
1  ALB18-A02-1  2018-08-30  North coast kelp emergent transects    ALB   
2  ALB18-A02-2  2018-08-30  North coast kelp emergent transects    ALB   
3   ALB18-A1-1  2018-08-30  North coast kelp emergent transects    ALB   
4   ALB18-A1-2  2018-08-30  North coast kelp emergent transects    ALB   

               locality  
0  Fort Ross State Park  
1            Albion Bay  
2            Albion Bay  
3            Albion Bay  
4            Albion Bay

In [11]:
## countryCode

event['countryCode'] = 'US'
event.head()

eventID   eventDate                            datasetID SiteID  \
0    FR18-D5-1  2018-07-12  North coast kelp emergent transects     FR   
1  ALB18-A02-1  2018-08-30  North coast kelp emergent transects    ALB   
2  ALB18-A02-2  2018-08-30  North coast kelp emergent transects    ALB   
3   ALB18-A1-1  2018-08-30  North coast kelp emergent transects    ALB   
4   ALB18-A1-2  2018-08-30  North coast kelp emergent transects    ALB   

               locality countryCode  
0  Fort Ross State Park          US  
1            Albion Bay          US  
2            Albion Bay          US  
3            Albion Bay          US  
4            Albion Bay          US

## START HERE
1. This dataset is unusual in that there are individual coordinates for each transect.
2. Just add them in, and don't worry about a site table for now.
3. Although maybe we'll still need one for DataONE?

**The site table doesn't contain lat/lon information, and the survey table seems to contain weird forms of it.** I received the following information from Laura in June:

| Site           | ID  | Latitude  | Longitude  |
|----------------|-----|-----------|------------|
| Albion Bay     | AB  | 39 22.900 | 123 77.881 |
| Caspar Cove    | CC  | 39 36.556 | 123 82.362 |
| Fort Ross      | FR  | 38 50.797 | 123 23.851 |
| Ocean Cove     | OC  | 38 55.363 | 123 30.688 |
| Point Arena    | PA  | 38 54.783 | 123 42.835 |
| Point Cabrillo | PC  | 39 20.752 | 123 49.628 |
| Russian Gulch  | RG  | 39 32.687 | 123 80.782 |
| Salt Point     | SP  | 38 33.972 | 123 20.172 |
| Sea Ranch      | SER | 38 42.468 | 123 27.302 |
| Timber Cove    | TC  | 38 53.476 | 123 28.192 |
| Todds Point    | TP  | 38 43.195 | 123 81.780 |
| Van Damme      | VD  | 39 16.044 | 123 47.662 |

**Note** that something is clearly wrong here. I assume these coordinates are in degrees + decimal minutes, but several of the decimal minute entries are > 60. This shouldn't be.

In [11]:
## Create a digital copy of the table Laura sent

# Define columns
site_name = ['Albion Bay', 'Caspar Cove', 'Fort Ross', 'Ocean Cove', 'Point Arena', 'Point Cabrillo', 'Russian Gulch', 'Salt Point', 'Sea Ranch', 'Timber Cove', 'Todds Point', 'Van Damme']
site_id = ['AB', 'CC', 'FR', 'OC', 'PA', 'PC', 'RG', 'SP', 'SER', 'TC', 'TP', 'VD']
lat_deg = [39, 39, 38, 38, 38, 39, 39, 38, 38, 38, 38, 39]
lat_min = [22.900, 36.556, 50.797, 55.363, 54.783, 20.752, 32.687, 33.972, 42.468, 53.476, 43.195, 16.044]
lon_deg = [123]*12
lon_min = [77.881, 82.362, 23.851, 30.688, 42.835, 49.628, 80.782, 20.172, 27.302, 28.192, 81.780, 47.662]

# Create df
site_lat_lon = pd.DataFrame({'Site':site_name,
                            'ID':site_id,
                            'Lat_deg':lat_deg,
                            'Lat_min':lat_min,
                            'Lon_deg':lon_deg,
                            'Lon_min':lon_min})

# Convert lat, lons to decimal degrees
site_lat_lon['Lat_dd'] = site_lat_lon['Lat_deg'] + site_lat_lon['Lat_min']/60
site_lat_lon['Lon_dd'] = site_lat_lon['Lon_deg'] + site_lat_lon['Lon_min']/60

# Merge with site table to populate what few lat, lons we know at this time. **Note that based on the prior text box, some of these are wrong.**
site = site.merge(site_lat_lon, how='left', left_on='SiteID', right_on='ID')
site.head()

SiteID                                      SITE         Site   ID  Lat_deg  \
0    ALB                                Albion Bay          NaN  NaN      NaN   
1     BR           Bodega Marine Life Refuge (BML)          NaN  NaN      NaN   
2    CAT  Catalina Island, Bird Rock (Southern CA)          NaN  NaN      NaN   
3     CC                               Caspar Cove  Caspar Cove   CC     39.0   
4     FM                            Fisk Mill Cove          NaN  NaN      NaN   

   Lat_min  Lon_deg  Lon_min     Lat_dd    Lon_dd  
0      NaN      NaN      NaN        NaN       NaN  
1      NaN      NaN      NaN        NaN       NaN  
2      NaN      NaN      NaN        NaN       NaN  
3   36.556    123.0   82.362  39.609267  124.3727  
4      NaN      NaN      NaN        NaN       NaN

In [12]:
## decimalLatitude, decimalLongitude

occ = occ.merge(site[['SiteID', 'Lat_dd', 'Lon_dd']], how='left', on='SiteID')
occ.rename(columns={'Lat_dd':'decimalLatitude', 'Lon_dd':'decimalLongitude'}, inplace=True)
print(occ.shape)
occ.head()

(455, 8)


eventID   eventDate                            datasetID SiteID  \
0  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
1  BR99-02  1999-07-29  North coast kelp emergent transects     BR   
2  BR99-03  1999-07-29  North coast kelp emergent transects     BR   
3  BR99-04  1999-07-29  North coast kelp emergent transects     BR   
4  BR99-05  1999-07-29  North coast kelp emergent transects     BR   

                          locality countryCode  decimalLatitude  \
0  Bodega Marine Life Refuge (BML)          US              NaN   
1  Bodega Marine Life Refuge (BML)          US              NaN   
2  Bodega Marine Life Refuge (BML)          US              NaN   
3  Bodega Marine Life Refuge (BML)          US              NaN   
4  Bodega Marine Life Refuge (BML)          US              NaN   

   decimalLongitude  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4               NaN

**Note** that once merged, the only lat, lons that are missing are from Bodega Marine Life Refuge.

In [13]:
## coordinateUncertaintyInMeters

occ['coordinateUncertaintyInMeters'] = 250
occ.head()

eventID   eventDate                            datasetID SiteID  \
0  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
1  BR99-02  1999-07-29  North coast kelp emergent transects     BR   
2  BR99-03  1999-07-29  North coast kelp emergent transects     BR   
3  BR99-04  1999-07-29  North coast kelp emergent transects     BR   
4  BR99-05  1999-07-29  North coast kelp emergent transects     BR   

                          locality countryCode  decimalLatitude  \
0  Bodega Marine Life Refuge (BML)          US              NaN   
1  Bodega Marine Life Refuge (BML)          US              NaN   
2  Bodega Marine Life Refuge (BML)          US              NaN   
3  Bodega Marine Life Refuge (BML)          US              NaN   
4  Bodega Marine Life Refuge (BML)          US              NaN   

   decimalLongitude  coordinateUncertaintyInMeters  
0               NaN                            250  
1               NaN                            250  
2               NaN                            250  
3               NaN                            250  
4               NaN                            250

**Is 250 m reasonable for coordinateUncertaintyInMeters?**

In [14]:
## occurrenceID

# Merge with counts
occ = occ.merge(count[['Survey_Num', 'SpeciesID', 'COUNT']], how='left', left_on='eventID', right_on='Survey_Num')

# Create occurrenceID
occ['occurrenceID'] = occ.groupby('eventID').cumcount()+1
occ['occurrenceID'] = occ['eventID'] + '_occ' + occ['occurrenceID'].astype(str)

# Drop redundant columns
occ.drop(columns=['Survey_Num'], inplace=True)

print(occ.shape)
occ.head()

(3440, 12)


eventID   eventDate                            datasetID SiteID  \
0  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
1  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
2  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
3  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
4  BR99-01  1999-07-29  North coast kelp emergent transects     BR   

                          locality countryCode  decimalLatitude  \
0  Bodega Marine Life Refuge (BML)          US              NaN   
1  Bodega Marine Life Refuge (BML)          US              NaN   
2  Bodega Marine Life Refuge (BML)          US              NaN   
3  Bodega Marine Life Refuge (BML)          US              NaN   
4  Bodega Marine Life Refuge (BML)          US              NaN   

   decimalLongitude  coordinateUncertaintyInMeters SpeciesID COUNT  \
0               NaN                            250        A1     0   
1               NaN                            250       A11     0   
2               NaN                            250       A12     0   
3               NaN                            250        C1     0   
4               NaN                            250       F11     0   

   occurrenceID  
0  BR99-01_occ1  
1  BR99-01_occ2  
2  BR99-01_occ3  
3  BR99-01_occ4  
4  BR99-01_occ5

In [15]:
## scientificName

# Merge to get species names from codes
occ = occ.merge(species[['SpeciesID', 'SPECIES', 'Scientific']], how='left', on='SpeciesID')

# Rename columns
occ.rename(columns={'SPECIES':'vernacularName',
                   'Scientific':'scientificName'}, inplace=True)

# Drop
occ.drop(columns=['SpeciesID'], inplace=True)

print(occ.shape)
occ.head()

(3440, 13)


eventID   eventDate                            datasetID SiteID  \
0  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
1  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
2  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
3  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
4  BR99-01  1999-07-29  North coast kelp emergent transects     BR   

                          locality countryCode  decimalLatitude  \
0  Bodega Marine Life Refuge (BML)          US              NaN   
1  Bodega Marine Life Refuge (BML)          US              NaN   
2  Bodega Marine Life Refuge (BML)          US              NaN   
3  Bodega Marine Life Refuge (BML)          US              NaN   
4  Bodega Marine Life Refuge (BML)          US              NaN   

   decimalLongitude  coordinateUncertaintyInMeters COUNT  occurrenceID  \
0               NaN                            250     0  BR99-01_occ1   
1               NaN                            250     0  BR99-01_occ2   
2               NaN                            250     0  BR99-01_occ3   
3               NaN                            250     0  BR99-01_occ4   
4               NaN                            250     0  BR99-01_occ5   

  vernacularName              scientificName  
0    Red Abalone                H. Rufescens  
1   Flat Abalone              H. walallensis  
2  Pinto Abalone            H. kamtschatkana  
3     Cancer sp.                     general  
4        Cabezon  Scorpaenichthys marmoratus

In [16]:
## Get unique names

names = occ['scientificName'].unique()
names

array(['H. Rufescens', 'H. walallensis', 'H. kamtschatkana', 'general',
       'Scorpaenichthys marmoratus', 'Ophiodon elongatus',
       'Pisaster spp.', 'Dermasterias imbricata', 'Henricia leviuscula',
       'Patiria miniata', 'Othasterias koehleri',
       'Pycnopodia helianthoides', 'S. franciscanus', 'S. purpuratus'],
      dtype=object)

**There are a number of changes that need to be made to scientific names before they can go through WoRMS.**
1. A number of abalone species don't have the genus written out
    - H. Rufescens, H. walallensis, H. kamtschatkana --> Haliotis
2. A number of urchin species don't have the genus written out
    - S. franciscanus and S. purpuratus --> Strongylocentrotus
3. A number of scientific names are simply "general"
    - These seem to correspond to vernacular names: Cancer sp., Solaster species
    
**In addition,** misspellings:
- Othasterias koehleri should be Orthasterias koehleri

In [17]:
## Update names and occ based on the above observations

# Update occ
occ.replace({'H. Rufescens':'Haliotis rufescens',
            'H. walallensis':'Haliotis walallensis',
            'H. kamtschatkana':'Haliotis kamtschatkana',
            'S. franciscanus':'Strongylocentrotus franciscanus',
            'S. purpuratus':'Strongylocentrotus purpuratus',
            'Othasterias koehleri':'Orthasterias koehleri'}, inplace=True)
occ.loc[occ['vernacularName'] == 'Cancer sp.', 'scientificName'] = 'Cancer sp.'
occ.loc[occ['vernacularName'] == 'Solaster species', 'scientificName'] = 'Solaster species'

# Get unique names again
names = occ['scientificName'].unique()
names

array(['Haliotis rufescens', 'Haliotis walallensis',
       'Haliotis kamtschatkana', 'Cancer sp.',
       'Scorpaenichthys marmoratus', 'Ophiodon elongatus',
       'Pisaster spp.', 'Solaster species', 'Dermasterias imbricata',
       'Henricia leviuscula', 'Patiria miniata', 'Orthasterias koehleri',
       'Pycnopodia helianthoides', 'Strongylocentrotus franciscanus',
       'Strongylocentrotus purpuratus'], dtype=object)

In [18]:
## Run through WoRMS

name_id_dict, name_name_dict, name_taxid_dict, name_class_dict = WoRMS.run_get_worms_from_scientific_name(names, verbose_flag=True)

Url didn't work for Cancer sp. checking:  Cancer
Url didn't work for Pisaster spp. checking:  Pisaster
Url didn't work for Solaster species checking:  Solaster


In [19]:
## Add scientific name-related columns

occ['scientificNameID'] = occ['scientificName']
occ['scientificNameID'].replace(name_id_dict, inplace=True)

occ['taxonID'] = occ['scientificName']
occ['taxonID'].replace(name_taxid_dict, inplace=True)
occ.head()

eventID   eventDate                            datasetID SiteID  \
0  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
1  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
2  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
3  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
4  BR99-01  1999-07-29  North coast kelp emergent transects     BR   

                          locality countryCode  decimalLatitude  \
0  Bodega Marine Life Refuge (BML)          US              NaN   
1  Bodega Marine Life Refuge (BML)          US              NaN   
2  Bodega Marine Life Refuge (BML)          US              NaN   
3  Bodega Marine Life Refuge (BML)          US              NaN   
4  Bodega Marine Life Refuge (BML)          US              NaN   

   decimalLongitude  coordinateUncertaintyInMeters COUNT  occurrenceID  \
0               NaN                            250     0  BR99-01_occ1   
1               NaN                            250     0  BR99-01_occ2   
2               NaN                            250     0  BR99-01_occ3   
3               NaN                            250     0  BR99-01_occ4   
4               NaN                            250     0  BR99-01_occ5   

  vernacularName              scientificName  \
0    Red Abalone          Haliotis rufescens   
1   Flat Abalone        Haliotis walallensis   
2  Pinto Abalone      Haliotis kamtschatkana   
3     Cancer sp.                  Cancer sp.   
4        Cabezon  Scorpaenichthys marmoratus   

                            scientificNameID  taxonID  
0  urn:lsid:marinespecies.org:taxname:445357   445357  
1  urn:lsid:marinespecies.org:taxname:445374   445374  
2  urn:lsid:marinespecies.org:taxname:405014   405014  
3  urn:lsid:marinespecies.org:taxname:106876   106876  
4  urn:lsid:marinespecies.org:taxname:282726   282726

In [20]:
## Since there are no identificationQualifiers so far, swap out scientificName too

occ['scientificName'].replace(name_name_dict, inplace=True)
occ.head()

eventID   eventDate                            datasetID SiteID  \
0  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
1  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
2  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
3  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
4  BR99-01  1999-07-29  North coast kelp emergent transects     BR   

                          locality countryCode  decimalLatitude  \
0  Bodega Marine Life Refuge (BML)          US              NaN   
1  Bodega Marine Life Refuge (BML)          US              NaN   
2  Bodega Marine Life Refuge (BML)          US              NaN   
3  Bodega Marine Life Refuge (BML)          US              NaN   
4  Bodega Marine Life Refuge (BML)          US              NaN   

   decimalLongitude  coordinateUncertaintyInMeters COUNT  occurrenceID  \
0               NaN                            250     0  BR99-01_occ1   
1               NaN                            250     0  BR99-01_occ2   
2               NaN                            250     0  BR99-01_occ3   
3               NaN                            250     0  BR99-01_occ4   
4               NaN                            250     0  BR99-01_occ5   

  vernacularName              scientificName  \
0    Red Abalone          Haliotis rufescens   
1   Flat Abalone        Haliotis walallensis   
2  Pinto Abalone      Haliotis kamtschatkana   
3     Cancer sp.                      Cancer   
4        Cabezon  Scorpaenichthys marmoratus   

                            scientificNameID  taxonID  
0  urn:lsid:marinespecies.org:taxname:445357   445357  
1  urn:lsid:marinespecies.org:taxname:445374   445374  
2  urn:lsid:marinespecies.org:taxname:405014   405014  
3  urn:lsid:marinespecies.org:taxname:106876   106876  
4  urn:lsid:marinespecies.org:taxname:282726   282726

In [21]:
## Add other name-related columns

occ['nameAccordingTo'] = 'WoRMS'
occ['occurrenceStatus'] = 'present'
occ['basisOfRecord'] = 'HumanObservation'

occ.head()

eventID   eventDate                            datasetID SiteID  \
0  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
1  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
2  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
3  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
4  BR99-01  1999-07-29  North coast kelp emergent transects     BR   

                          locality countryCode  decimalLatitude  \
0  Bodega Marine Life Refuge (BML)          US              NaN   
1  Bodega Marine Life Refuge (BML)          US              NaN   
2  Bodega Marine Life Refuge (BML)          US              NaN   
3  Bodega Marine Life Refuge (BML)          US              NaN   
4  Bodega Marine Life Refuge (BML)          US              NaN   

   decimalLongitude  coordinateUncertaintyInMeters COUNT  occurrenceID  \
0               NaN                            250     0  BR99-01_occ1   
1               NaN                            250     0  BR99-01_occ2   
2               NaN                            250     0  BR99-01_occ3   
3               NaN                            250     0  BR99-01_occ4   
4               NaN                            250     0  BR99-01_occ5   

  vernacularName              scientificName  \
0    Red Abalone          Haliotis rufescens   
1   Flat Abalone        Haliotis walallensis   
2  Pinto Abalone      Haliotis kamtschatkana   
3     Cancer sp.                      Cancer   
4        Cabezon  Scorpaenichthys marmoratus   

                            scientificNameID  taxonID nameAccordingTo  \
0  urn:lsid:marinespecies.org:taxname:445357   445357           WoRMS   
1  urn:lsid:marinespecies.org:taxname:445374   445374           WoRMS   
2  urn:lsid:marinespecies.org:taxname:405014   405014           WoRMS   
3  urn:lsid:marinespecies.org:taxname:106876   106876           WoRMS   
4  urn:lsid:marinespecies.org:taxname:282726   282726           WoRMS   

  occurrenceStatus     basisOfRecord  
0          present  HumanObservation  
1          present  HumanObservation  
2          present  HumanObservation  
3          present  HumanObservation  
4          present  HumanObservation

In [22]:
## Get count information in the right place

# Set organismQuantity
occ['organismQuantity'] = occ['COUNT'].copy()

# Drop COUNT column
occ.drop(columns=['COUNT'], inplace=True)

# Add organismQuantityType
occ['organismQuantityType'] = 'density in number of individuals per 60m2'
occ.head()

eventID   eventDate                            datasetID SiteID  \
0  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
1  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
2  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
3  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
4  BR99-01  1999-07-29  North coast kelp emergent transects     BR   

                          locality countryCode  decimalLatitude  \
0  Bodega Marine Life Refuge (BML)          US              NaN   
1  Bodega Marine Life Refuge (BML)          US              NaN   
2  Bodega Marine Life Refuge (BML)          US              NaN   
3  Bodega Marine Life Refuge (BML)          US              NaN   
4  Bodega Marine Life Refuge (BML)          US              NaN   

   decimalLongitude  coordinateUncertaintyInMeters  occurrenceID  \
0               NaN                            250  BR99-01_occ1   
1               NaN                            250  BR99-01_occ2   
2               NaN                            250  BR99-01_occ3   
3               NaN                            250  BR99-01_occ4   
4               NaN                            250  BR99-01_occ5   

  vernacularName              scientificName  \
0    Red Abalone          Haliotis rufescens   
1   Flat Abalone        Haliotis walallensis   
2  Pinto Abalone      Haliotis kamtschatkana   
3     Cancer sp.                      Cancer   
4        Cabezon  Scorpaenichthys marmoratus   

                            scientificNameID  taxonID nameAccordingTo  \
0  urn:lsid:marinespecies.org:taxname:445357   445357           WoRMS   
1  urn:lsid:marinespecies.org:taxname:445374   445374           WoRMS   
2  urn:lsid:marinespecies.org:taxname:405014   405014           WoRMS   
3  urn:lsid:marinespecies.org:taxname:106876   106876           WoRMS   
4  urn:lsid:marinespecies.org:taxname:282726   282726           WoRMS   

  occurrenceStatus     basisOfRecord organismQuantity  \
0          present  HumanObservation                0   
1          present  HumanObservation                0   
2          present  HumanObservation                0   
3          present  HumanObservation                0   
4          present  HumanObservation                0   

                        organismQuantityType  
0  density in number of individuals per 60m2  
1  density in number of individuals per 60m2  
2  density in number of individuals per 60m2  
3  density in number of individuals per 60m2  
4  density in number of individuals per 60m2

**Note** that I should check the organismQuantityType values.

**Also note** that organismQuantity is currently a string. I think this is because there were weird values in the original data set. These have been eliminated during the EMERGENT filtering step, but may return once that filter's been broadened, which I think it needs to be. **Also, 274 rows in the count table have COUNT = NaN.** I'm not sure if this is meaningful, or if it's an error. Because of the merge, these rows also appear in occ, and I'm going to leave them for now. But I'll have to make organismQuantity a float for now rather than an int.

In [23]:
## Update presence/absence status

# Ensure organismQuantity is an int
occ['organismQuantity'] = pd.to_numeric(occ['organismQuantity'])

# Change 0 counts to absent
occ.loc[occ['organismQuantity'] == 0, 'occurrenceStatus'] = 'absent'
occ.head()

eventID   eventDate                            datasetID SiteID  \
0  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
1  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
2  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
3  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
4  BR99-01  1999-07-29  North coast kelp emergent transects     BR   

                          locality countryCode  decimalLatitude  \
0  Bodega Marine Life Refuge (BML)          US              NaN   
1  Bodega Marine Life Refuge (BML)          US              NaN   
2  Bodega Marine Life Refuge (BML)          US              NaN   
3  Bodega Marine Life Refuge (BML)          US              NaN   
4  Bodega Marine Life Refuge (BML)          US              NaN   

   decimalLongitude  coordinateUncertaintyInMeters  occurrenceID  \
0               NaN                            250  BR99-01_occ1   
1               NaN                            250  BR99-01_occ2   
2               NaN                            250  BR99-01_occ3   
3               NaN                            250  BR99-01_occ4   
4               NaN                            250  BR99-01_occ5   

  vernacularName              scientificName  \
0    Red Abalone          Haliotis rufescens   
1   Flat Abalone        Haliotis walallensis   
2  Pinto Abalone      Haliotis kamtschatkana   
3     Cancer sp.                      Cancer   
4        Cabezon  Scorpaenichthys marmoratus   

                            scientificNameID  taxonID nameAccordingTo  \
0  urn:lsid:marinespecies.org:taxname:445357   445357           WoRMS   
1  urn:lsid:marinespecies.org:taxname:445374   445374           WoRMS   
2  urn:lsid:marinespecies.org:taxname:405014   405014           WoRMS   
3  urn:lsid:marinespecies.org:taxname:106876   106876           WoRMS   
4  urn:lsid:marinespecies.org:taxname:282726   282726           WoRMS   

  occurrenceStatus     basisOfRecord  organismQuantity  \
0           absent  HumanObservation               0.0   
1           absent  HumanObservation               0.0   
2           absent  HumanObservation               0.0   
3           absent  HumanObservation               0.0   
4           absent  HumanObservation               0.0   

                        organismQuantityType  
0  density in number of individuals per 60m2  
1  density in number of individuals per 60m2  
2  density in number of individuals per 60m2  
3  density in number of individuals per 60m2  
4  density in number of individuals per 60m2

In [24]:
## Depth

occ = occ.merge(survey[['Survey_Num', 'Min_DEPTH', 'Max_Depth']], how='left', left_on='eventID', right_on='Survey_Num')
occ.drop(columns=['Survey_Num'], inplace=True)
occ.rename(columns={'Min_DEPTH':'minimumDepthInMeters',
           'Max_Depth':'maximumDepthInMeters'}, inplace=True)
print(occ.shape)
occ.head()

(3440, 21)


eventID   eventDate                            datasetID SiteID  \
0  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
1  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
2  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
3  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
4  BR99-01  1999-07-29  North coast kelp emergent transects     BR   

                          locality countryCode  decimalLatitude  \
0  Bodega Marine Life Refuge (BML)          US              NaN   
1  Bodega Marine Life Refuge (BML)          US              NaN   
2  Bodega Marine Life Refuge (BML)          US              NaN   
3  Bodega Marine Life Refuge (BML)          US              NaN   
4  Bodega Marine Life Refuge (BML)          US              NaN   

   decimalLongitude  coordinateUncertaintyInMeters  occurrenceID  ...  \
0               NaN                            250  BR99-01_occ1  ...   
1               NaN                            250  BR99-01_occ2  ...   
2               NaN                            250  BR99-01_occ3  ...   
3               NaN                            250  BR99-01_occ4  ...   
4               NaN                            250  BR99-01_occ5  ...   

               scientificName                           scientificNameID  \
0          Haliotis rufescens  urn:lsid:marinespecies.org:taxname:445357   
1        Haliotis walallensis  urn:lsid:marinespecies.org:taxname:445374   
2      Haliotis kamtschatkana  urn:lsid:marinespecies.org:taxname:405014   
3                      Cancer  urn:lsid:marinespecies.org:taxname:106876   
4  Scorpaenichthys marmoratus  urn:lsid:marinespecies.org:taxname:282726   

  taxonID  nameAccordingTo occurrenceStatus     basisOfRecord  \
0  445357            WoRMS           absent  HumanObservation   
1  445374            WoRMS           absent  HumanObservation   
2  405014            WoRMS           absent  HumanObservation   
3  106876            WoRMS           absent  HumanObservation   
4  282726            WoRMS           absent  HumanObservation   

  organismQuantity                       organismQuantityType  \
0              0.0  density in number of individuals per 60m2   
1              0.0  density in number of individuals per 60m2   
2              0.0  density in number of individuals per 60m2   
3              0.0  density in number of individuals per 60m2   
4              0.0  density in number of individuals per 60m2   

  minimumDepthInMeters  maximumDepthInMeters  
0                 39.0                   NaN  
1                 39.0                   NaN  
2                 39.0                   NaN  
3                 39.0                   NaN  
4                 39.0                   NaN  

[5 rows x 21 columns]

**Note** that when maximumDepthInMeters is not NaN, it is equal to minimumDepthInMeters. These values look like they're in feet, but **I should check this.** Also, **when was average depth recorded**, and how should I include it?

In [191]:
## samplingProtocol, samplingEffort

occ['samplingProtocol'] = 'band transect'
occ['samplingEffort'] = '10-15 minutes per transect'
occ.head()

eventID   eventDate                            datasetID SiteID  \
0  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
1  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
2  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
3  BR99-01  1999-07-29  North coast kelp emergent transects     BR   
4  BR99-01  1999-07-29  North coast kelp emergent transects     BR   

                          locality countryCode  decimalLatitude  \
0  Bodega Marine Life Refuge (BML)          US              NaN   
1  Bodega Marine Life Refuge (BML)          US              NaN   
2  Bodega Marine Life Refuge (BML)          US              NaN   
3  Bodega Marine Life Refuge (BML)          US              NaN   
4  Bodega Marine Life Refuge (BML)          US              NaN   

   decimalLongitude  coordinateUncertaintyInMeters  occurrenceID  ... taxonID  \
0               NaN                            250  BR99-01_occ1  ...  445357   
1               NaN                            250  BR99-01_occ2  ...  445374   
2               NaN                            250  BR99-01_occ3  ...  405014   
3               NaN                            250  BR99-01_occ4  ...  106876   
4               NaN                            250  BR99-01_occ5  ...  282726   

  nameAccordingTo occurrenceStatus     basisOfRecord organismQuantity  \
0           WoRMS           absent  HumanObservation              0.0   
1           WoRMS           absent  HumanObservation              0.0   
2           WoRMS           absent  HumanObservation              0.0   
3           WoRMS           absent  HumanObservation              0.0   
4           WoRMS           absent  HumanObservation              0.0   

                        organismQuantityType minimumDepthInMeters  \
0  density in number of individuals per 60m2                 39.0   
1  density in number of individuals per 60m2                 39.0   
2  density in number of individuals per 60m2                 39.0   
3  density in number of individuals per 60m2                 39.0   
4  density in number of individuals per 60m2                 39.0   

   maximumDepthInMeters samplingProtocol              samplingEffort  
0                   NaN    band transect  10-15 minutes per transect  
1                   NaN    band transect  10-15 minutes per transect  
2                   NaN    band transect  10-15 minutes per transect  
3                   NaN    band transect  10-15 minutes per transect  
4                   NaN    band transect  10-15 minutes per transect  

[5 rows x 23 columns]

**Check samplingProtocol and samplingEffort.**

In [192]:
## Save

occ.to_csv('occ_20201028.csv', index=False, na_rep='NaN')

## Questions

1. If you want habitat and size data to be included, where are these located? **The size data should be in Tbl_new_size, and the substrate data should be in tblSubstrate. tblHabitat should contain algae data.**
2. Laura said that she only wanted to share the data from emergent surveys. If I filter for TYPE = 'EMERGENT', I end up discarding the majority of the data (all but 455 records out of ~ 3400). Is this really the right thing to do? There are the following values under SURVEY, many of which use the word 'Emergent'. **Laura now says she would like to share all of the data (including the Rapid emergent, or REAS surveys). Below, I've indicated Y or N for whether Laura thought the survey type should be included. She also said that anything with 'emergent' somewhere in the title should have TYPE = EMERGENT.**
- 'Transect - 30m (Rapid Emergent)' - Y
- 'Transect - 20mx2m (Emergent)' - N
- nan
- 'ARM' - N
- 'Transect - 30mx2m (Emergent)' - Y (old surveys?)
- 'Swim' - N
- 'Transect - 5m (Invasive)' - N
- 'Transect - 20mx5m (Emergent)' - N
- 'Transect - 16mx2m (Emergent)' - N
- 'Transect - 1/4m' - N
- 'Transect - 1/4M' - N
- 'Transect - running transect' - N
- 'Transect - Running Transect' - N
- 'Transect - 30m (Emergent)' (this is the only value left after filtering by TYPE = EMERGENT) - Y
3. Is it possible to get a general description of what happens during the emergent surveys? How many sites are surveyed and how often? Are some sites inside MPAs and some outside MPAs? Which? How many transects are done per site? Are quadrats important here? **Laura conducts a lot of transects at a small number of sites per year. Started in 1999 with only 1 site. Within each site, transects are depth-stratified into four categories: A (shallow 0-15 ft), B (16-30 ft), C (31-45 ft), D (46-60 ft). Within each strata, they randomly place 36 (or, at least, they try for 36, it's sometimes a little more and sometimes less) 3 x 2 m transects. The random placement occurs by randomly selecting GPS points. Over time, they have eliminated GPS points over sandy habitat; if a diver dropped on their coordinate, and it was > 50% sand, the transect was not conducted and the point was removed from future drawings. During each transect, there is a left and right diver, who each survey their respective 1 x 30 m area. Divers record inverts, benthic fish, substrate type, algae type, depth, and size. They take as long as they have to to survey everything they can find. To quantify substrate type, they stop at the 0, 10, 20, and 30 m marks, look at the 1 m2 area in front of them, and estimate the % of each substrate category. Algae type is recorded by functional group (e.g. crustose, etc.), and can add up to > 100% because they capture layering. Depth is recorded at the 0, 10, 20, and 30 m marks. Target species are sized during each transects (e.g. abalone, urchins); typically, if there are > 30 animals on a transect, only about 30 are sized. Numbers of organisms are recorded in 5 m bins - so for each organism, there should be 6 numbers for abundance on the right side of the transect, and 6 numbers for each organism on the left side of the transect. A few other points to keep in mind:**
- **Some sites (Salt Point, Casper) contain regions that were historically no red sea urchin fishing zones. They distinguish transects that are inside and outside of these zones (how?). Also, these sites often have more than 36 transects, because they try to do additional transects inside the historical no-take zones.**
- **Note that C and D depth strata are out of reach for breath-hold abalone divers, and so are effectively protected**

**Rapid Emergent surveys are slightly different and are conducted in years when they want to survey more sites quickly. A dive team drops in a single location, and runs four transects from that location (so these transects are not as spatially independent). Otherwise, the transects are identical as described above, except that substrate is quantified using UPC at every meter along the transect tape.**

4. How are entries in Survey_Num formatted? Most look like the site code + the year + a number. Is the number a transect number? There are a number of Survey_Nums that do not fit this format, for example:
- 'PA71C-01'
- 'PA86dfg-13'

**Site code contains a unique 2 letter code (Point Arena = PA, etc.) + a 2 digit year (99 - 19) + transect information. Transects are usually accompanied by a letter indicating the depth strata (e.g. A02). If the dive team had to move their transect off of the designated coordinate for some reason, it's indicated by AA (e.g. BAA7 indicates an alternate transect for B7). There are also data from 1971 and 1986, 1971. These are historical data from surveys that Laura did not conduct. The methods were similar, although the sites were all accessible from shore. These survey numbers may be weird. There is a report somewhere on these data that may help us flesh out differences in survey methods.** 

5. By filtering by TYPE = EMERGENT, we are also only getting survey dates between 1971 and 2001. This doesn't seem right.
6. What are the correct lat, lons for the survey sites? Columns in the survey table are confusing. In June, Laura sent me a screenshot of another table, but the lat, lons are confusing there, too. Specifically, they appear to be in degrees plus decimal minutes format, but several of the decimal minutes values are > 60. (E.g. Albion Bay, 39 22.900, 123 77.881) **Laura sees that there's something wrong here; she will look for better coordinates. Unlike the other surveys, she also does have transect-level coordinates, which she would prefer to include if possible. I said I'd look at the data and try to give her a better picture of where transect-level coordinates are missing.**
7. Is 250 m reasonable for coordinateUncertaintyInMeters?
8. There are 274 values in the count table where COUNT = NaN. What does this mean? Are these records different from when COUNT = 0? Also, there are some other weird values in the COUNT column, including:
- 'RG18-C3-'
- 'P'
9. How and when was depth recorded for these surveys? Should I use average, min, max?
10. What should I put for samplingProtocol? samplingEffort? **samplingProtocol = band transect (I didn't check but this is likely OK. Or maybe she'd prefer rapid versus emergent). samplingEffort = survey everything.**

### In addition, Laura would like to cover:
1. The left and right side of the transects in the data base and possibly combining them. **Currently, the transect data for each species is in the form of 6 data points on the R and 6 on the left. We would have to sum them to get a total count for each transect. Should we keep it at 12 numbers or convert it to 1? This is a decision we should keep in mind and make eventually.**
2. The total number of transects and how they can change depending on the species **I'm a bit confused on this point, but Laura said that sometimes the number of transects for each species change at a site. E.g., sometimes there are 36 transects for abalone, and 10 for urchin. I didn't think transects were separated out by species, but perhaps this will become clearer as I look at the data. She said this happens primarily in recent years, so I can keep an eye out for that.**
3. The need for zeros in some of the species on some transects. **There are some species where they were looked for, but none were found, and a 0 was not entered. Instead, it's empty. Laura would like these zeros populated. She said it would involve figuring out how many transects happened at that site that year, and then going in and looking and making sure we have all of the zeros. She said that almost all species were looked for in almost all years, except purple urchin in 2017. If I can get the data clean enough, I may be able to do something like I did for PISCO to populate missing absence records.**


**In addition: Ask Laura for site area estimation table. She said there are shapefiles for all of the sites; she can she can connect me with Paulo Serpa, her GIS guy for those. They would be good to include in the DataONE submission.**

In [6]:
sur

Survey_ID SiteID   Survey_Num                           SURVEY  \
0          4790     FR    FR18-D5-1  Transect - 30m (Rapid Emergent)   
1          4919    ALB  ALB18-A02-1  Transect - 30m (Rapid Emergent)   
2          4920    ALB  ALB18-A02-2  Transect - 30m (Rapid Emergent)   
3          4915    ALB   ALB18-A1-1  Transect - 30m (Rapid Emergent)   
4          4916    ALB   ALB18-A1-2  Transect - 30m (Rapid Emergent)   
...         ...    ...          ...                              ...   
3789       1858     VD      VD99-45        Transect - 30m (Emergent)   
3790       1859     VD      VD99-46        Transect - 30m (Emergent)   
3791       1860     VD      VD99-47        Transect - 30m (Emergent)   
3792       1861     VD      VD99-48        Transect - 30m (Emergent)   
3793       4804     VD     VD-BA4-4        Transect - 30m (Emergent)   

                     DATE  Avg Depth  Min_DEPTH  Max_Depth      SLAT  \
0     2018-07-12 00:00:00        NaN       48.0        NaN       NaN   
1     2018-08-30 00:00:00        NaN        7.0       10.0       NaN   
2     2018-08-30 00:00:00        NaN       10.0       14.0       NaN   
3     2018-08-30 00:00:00        NaN       12.0       16.0       NaN   
4     2018-08-30 00:00:00        NaN        8.0        8.0       NaN   
...                   ...        ...        ...        ...       ...   
3789  1999-08-25 00:00:00        0.0       25.0        NaN  391684.0   
3790  1999-08-25 00:00:00        0.0        8.0        NaN       NaN   
3791  1999-08-25 00:00:00        0.0       20.0        NaN       NaN   
3792  1999-08-04 00:00:00        0.0       45.0        NaN       NaN   
3793  2018-08-16 00:00:00        NaN       18.0       20.0       NaN   

          SLONG      ELAT  ELONG  SLAT_DD  SLONG_DD  ELAT_DD  ELONG_DD  \
0           NaN       NaN    NaN      0.0       0.0      0.0       0.0   
1           NaN       NaN    NaN      0.0       0.0      0.0       0.0   
2           NaN       NaN    NaN      0.0       0.0      0.0       0.0   
3           NaN       NaN    NaN      0.0       0.0      0.0       0.0   
4           NaN       NaN    NaN      0.0       0.0      0.0       0.0   
...         ...       ...    ...      ...       ...      ...       ...   
3789  1234774.0  391684.0    NaN      NaN       NaN      NaN       NaN   
3790        NaN       NaN    NaN      NaN       NaN      NaN       NaN   
3791        NaN       NaN    NaN      NaN       NaN      NaN       NaN   
3792        NaN       NaN    NaN      NaN       NaN      NaN       NaN   
3793        NaN       NaN    NaN      0.0       0.0      0.0       0.0   

                                               COMMENTS  
0     RAPID EMERGENT SURVEY. D5: 1 OUT OF 2 TRANSECT...  
1     RAPID EMERGENT SITE A02: 1 OF 2 TRANSECTS. HEA...  
2     RAPID EMERGENT SITE A02: 2 OF 2 TRANSECTS. HEA...  
3     RAPID EMERGENT SITE A1: 1 OF 4 TRANSECTS. HEAD...  
4     RAPID EMERGENT SITE A1: 2 OF 4 TRANSECTS. HEAD...  
...                                                 ...  
3789                                                NaN  
3790                                                NaN  
3791                                                NaN  
3792                                                NaN  
3793  RAPID EMERGENT SURVEY. BA4: 4 OUT OF 4 TRANSEC...  

[2943 rows x 17 columns]